src/js/nav/globalNav.js

```diff
@@ -1,47 +1,63 @@
+    }), {});
 };
 
+export async function calculateVisibility({ permissions: appPermisions }, { permissions, id }, groupVisibility) {
+    const visibility = (permissions && visibilityFunctions[permissions.method]) ?
+        await visibilityFunctions[permissions.method](...permissions.args || []) :
+        true;
+    return (
+        isVisible([id], id, visibility) &&
+        isVisible(appPermisions && appPermisions.apps, id, groupVisibility)
+    );
+}
+
 // Returns a list of routes/subItems owned by an app
-function getRoutesForApp(app, masterConfig) {
+async function getRoutesForApp(app, masterConfig) {
     if (Object.prototype.hasOwnProperty.call(app, 'frontend') &&
         Object.prototype.hasOwnProperty.call(app.frontend, 'sub_apps')
     ) {
-        let routes = [];
-        app.frontend.sub_apps.forEach((subItem => {
-            let subAppData;
-            if (subItem.title) {
-                subAppData = {
+        const visibility = (app.permissions && visibilityFunctions[app.permissions.method]) ?
+            await visibilityFunctions[app.permissions.method](...app.permissions.args || []) :
+            true;
+
+        const routes = await Promise.all(app.frontend.sub_apps.map(async subItem => {
+            return (await calculateVisibility(app, subItem, visibility)) && ({
+                ...subItem.title ? {
                     id: subItem.id || '',
                     title: subItem.title
-                };
-            } else {
-                subAppData = getAppData(subItem.id || subItem, 'subItems', masterConfig);
-            }
-
-            if (!(subAppData.disabled_on_prod && window.location.hostname === 'cloud.redhat.com')) {
-                routes.push({
-                    ...subAppData,
-                    ...subItem.default && { default: subItem.default },
-                    ...subItem.group && { group: subItem.group },
-                    ...subItem.reload && { reload: subItem.reload }
-                });
-            }
+                } : await getAppData(subItem.id || subItem, 'subItems', masterConfig),
+                ...subItem.default && { default: subItem.default },
+                ...subItem.group && { group: subItem.group },
+                ...subItem.reload && { reload: subItem.reload }
+            });
         }));
-        return routes;
+
+        return routes.filter(subAppData => {
+            return (subAppData && subAppData.title) && !(subAppData.disabled_on_prod && window.location.hostname === 'cloud.redhat.com');
+        });
     }
+
+    return [];
 }
 
 // Gets the app's data from the master config, if it exists
-function getAppData(appId, propName, masterConfig) {
+async function getAppData(appId, propName, masterConfig) {
     const app = masterConfig[appId];
     if (app && Object.prototype.hasOwnProperty.call(app, 'frontend')) {
-        const routes = getRoutesForApp(app, masterConfig);
+        if (app.permissions && !app.permissions.apps) {
```

Comment:
Is `!app.permissions.apps` this check necessary?

Reply:
Yes, this hides the entire section. So if dev does not specify apps it is ment to hide the entire firt level navigation.